File: rd-cpp/src/rd_framework_cpp/src/test/cases/RdSetTest.cpp
#include "RdFrameworkTestBase.h"
#include "impl/RdSet.h"

#include <gtest/gtest.h>

using vi = std::vector<int>;

using namespace rd;
using namespace test;

TEST_F(RdFrameworkTestBase, set_statics)
{
	int32_t id = 1;

	RdSet<int> server_set;
	RdSet<int> client_set;

	statics(server_set, id);
	statics(client_set, id);

	vi log;

	server_set.advise(serverLifetimeDef.lifetime, [&](AddRemove kind, int v) { log.push_back((kind == AddRemove::ADD) ? v : -v); });

	client_set.add(2);
	client_set.add(0);
	client_set.add(1);
	client_set.add(8);

	EXPECT_EQ(vi(), log);

	bindStatic(serverProtocol.get(), server_set, static_name);
	bindStatic(clientProtocol.get(), client_set, static_name);
	EXPECT_EQ((vi{2, 0, 1, 8}), log);

	client_set.remove(1);
	EXPECT_EQ((vi{2, 0, 1, 8, -1}), log);

	server_set.remove(1);
	client_set.remove(1);
	EXPECT_EQ((vi{2, 0, 1, 8, -1}), log);

	client_set.remove(2);
	EXPECT_EQ((vi{2, 0, 1, 8, -1, -2}), log);

	client_set.clear();
	EXPECT_EQ((vi{2, 0, 1, 8, -1, -2, -0, -8}), log);

	AfterTest();
}

TEST_F(RdFrameworkTestBase, set_move)
{
	RdSet<int> set1;
	RdSet<int> set2(std::move(set1));

	AfterTest();
}

TEST_F(RdFrameworkTestBase, set_iterator)
{
	RdSet<int> set;
	EXPECT_EQ(set.end(), set.rbegin().base());
	set.addAll({1, 2, 3});
	EXPECT_EQ(set.end(), set.rbegin().base());
}

File: rd-kt/rd-framework/src/commonTest/kotlin/com/jetbrains/rd/framework/test/cases/RdSetTest.kt
package com.jetbrains.rd.framework.test.cases

import com.jetbrains.rd.framework.base.static
import com.jetbrains.rd.framework.impl.RdSet
import com.jetbrains.rd.framework.test.util.RdFrameworkTestBase
import com.jetbrains.rd.util.reactive.AddRemove
import kotlin.test.Test
import kotlin.test.assertEquals


class RdSetTest : RdFrameworkTestBase() {
    @Test
    fun testStatic() {
        val id = 1
        val serverSet = RdSet<Int>().static(id)
        val clientSet = RdSet<Int>().static(id)

        val log = arrayListOf<Int>()
        serverSet.advise(serverLifetimeDef.lifetime) { kind, v -> log.add(if (kind == AddRemove.Add) v else -v) }

        clientSet.add(2)
        clientSet.add(0)
        clientSet.add(1)
        clientSet.add(8)
        clientSet.add(3)


        assertEquals(listOf<Int>(), log)

        serverProtocol.bindStatic(serverSet, "top")
        clientProtocol.bindStatic(clientSet, "top")
        assertEquals(listOf(2, 0, 1, 8, 3), log)

        clientSet.retainAll(listOf(8, 3, 0))
        assertEquals(listOf(2, 0, 1, 8, 3, -2, -1), log)

        serverSet.remove(1)
        clientSet.remove(1)
        assertEquals(listOf(2, 0, 1, 8, 3, -2, -1), log)

        clientSet.removeAll { it == 3 }
        assertEquals(listOf(2, 0, 1, 8, 3, -2, -1, -3), log)


        clientSet.clear()
        assertEquals(listOf(2, 0, 1, 8, 3, -2, -1, -3, -0, -8), log)
    }
}